File: src/modifiers/textContentType.ts
import { createModifier } from './createModifier';

/// Subset of `WKTextContentType` (watchOS's content-type hints) that maps
/// cleanly across platforms. watchOS does not expose the full
/// `UITextContentType` surface, so this is intentionally limited to the
/// common, broadly-available cases.
export type TextContentTypeKind =
  | 'username'
  | 'password'
  | 'newPassword'
  | 'oneTimeCode'
  | 'emailAddress'
  | 'telephoneNumber'
  | 'name'
  | 'URL';

export interface TextContentTypeParams {
  type: TextContentTypeKind;
}

/// SwiftUI `.textContentType(_:)`. On watchOS this takes a `WKTextContentType`,
/// which hints the system about the semantic meaning of the field's text
/// (autofill credentials, one-time codes, etc.). Pass a type string directly
/// or an object.
export function textContentType(
  type: TextContentTypeKind
): ReturnType<typeof createModifier>;
export function textContentType(
  params: TextContentTypeParams
): ReturnType<typeof createModifier>;
export function textContentType(arg: TextContentTypeKind | TextContentTypeParams) {
  if (typeof arg === 'string') {
    return createModifier('textContentType', { type: arg });
  }
  return createModifier('textContentType', arg);
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/TextContentType.swift
import SwiftUI
import WatchKit

/// SwiftUI `.textContentType(_:)`. On watchOS this takes a `WKTextContentType`
/// (from WatchKit), hinting the system about the field's semantic content for
/// autofill / QuickType. The modifier overload is watchOS 6+, well under the
/// deployment target.
///
/// LIMITED: watchOS does not expose the full `UITextContentType` surface, so
/// only the broadly-available cases below are mapped. An unrecognized `type`
/// leaves the view unchanged. Every mapped case exists at the watchOS 9
/// deployment target.
enum RNWTextContentTypeModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("textContentType") { view, params, _ in
            guard let type = parse(params.string("type")) else { return view }
            return AnyView(view.textContentType(type))
        }
    }

    private static func parse(_ s: String?) -> WKTextContentType? {
        switch s {
        case "username":        return .username
        case "password":        return .password
        case "newPassword":     return .newPassword
        case "oneTimeCode":     return .oneTimeCode
        case "emailAddress":    return .emailAddress
        case "telephoneNumber": return .telephoneNumber
        case "name":            return .name
        case "URL":             return .URL
        default:                return nil
        }
    }
}
